### PR TITLE
Incrementally update dependency graph for new cases

### PR DESCRIPTION
**Pull request by sharwell**
* Removing all references from a project
* Adding references to a project that did not have any

Builds on work from #40358
Supersedes #67758

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs [hunk 1]
```diff
@@ -128,6 +128,17 @@ internal ProjectDependencyGraph WithProjectReferences(ProjectId projectId, IRead
         {
             Contract.ThrowIfFalse(_projectIds.Contains(projectId));
 
+            if (!_referencesMap.ContainsKey(projectId))
+            {
+                // This project doesn't have any references currently, so we delegate to WithAddedProjectReferences
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs:133]: >WithAddedProjectReferences [](http://example.com/codeflow?start=89&length=26)

Is this supposed to be WithAdditionalProjectReferences?

**Comment by ToddGrun** [comment]
>             var reverseReferencesMap = new Dictionary<ProjectId, HashSet<ProjectId>>();

Would it be worth using a pooled dictionary here? I'm also curious on your thoughts whether it could be beneficial to have a MultiAdd that special cased HashSet to use a pool there too. (I see an ArrayBuilder version that uses a pool)

---
Refers to: src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs:208 in 75e2c29. [](commit_id = 75e2c296e71f053017cf1afb3e16c2922b8e7a78, deletion_comment = False)

**Comment by sharwell** [comment]
> Would it be worth using a pooled dictionary here? I'm also curious on your thoughts whether it could be beneficial to have a MultiAdd that special cased HashSet to use a pool there too. (I see an ArrayBuilder version that uses a pool)

It's too early to tell IMO. This change addresses the two most common cases for this call, so it's possible there's no significant overhead remaining in `ComputeReverseReferencesMap`.

**Comment by ToddGrun** [comment]
> It's too early to tell IMO. This change addresses the two most common cases for this call, so it's possible there's no significant overhead remaining in `ComputeReverseReferencesMap`.

Reasonable. I'll take another look at the profiles once this change makes it's way into the test runs.

**Review by ToddGrun — APPROVED** [review]

**Pull request merged**